Frameworks/Foundation/NSScanner.mm

```diff
@@ -575,118 +557,283 @@ - (BOOL)scanUpToString:(NSString*)string intoString:(NSString**)intoString {
-                result[resultLength++] = unicode;
-            } else {
-                break; // used to be "return NO";
-            }
+            break;
         }
     }
 
-    if (scanStarted) {
+    if (result.size() > 0) {
         if (stringp) {
-            *stringp = [NSString stringWithCharacters:result length:resultLength];
+            *stringp = [NSString stringWithCharacters:result.data() length:result.size()];
         }
-    }
 
-    IwFree(result);
+        _scanLocation = i;
+        return YES;
+    }
 
-    return scanStarted;
+    return NO;
 }
 
 /**
  @Status Interoperable
 */
-- (BOOL)scanUpToCharactersFromSet:(id)charset intoString:(NSString**)stringp {
-    unsigned int length = (unsigned int)[_string length];
-    unichar* result = (unichar*)IwMalloc(length * sizeof(unichar));
-    int resultLength = 0;
-    BOOL scanStarted = NO;
-    int oldLocation = _location;
-
-    for (; _location < length; _location++) {
-        unichar unicode = (unichar)[_string characterAtIndex:_location];
-
-        if ([_skipSet characterIsMember:unicode] && scanStarted == NO)
-            continue;
-        else if ([charset characterIsMember:unicode])
+- (BOOL)scanUpToCharactersFromSet:(NSCharacterSet* _Nonnull)charset intoString:(NSString* _Nullable*)stringp {
+    const NSUInteger length = _string.length;
+    NSUInteger i = [self _indexOfNextUnskippedCharacter];
+
+    std::vector<unichar> result;
+    result.reserve(length - i);
+    for (; i < length; ++i) {
+        unichar unicode = (unichar)[_string characterAtIndex:i];
+
+        if ([charset characterIsMember:unicode]) {
             break;
-        else {
-            scanStarted = YES;
-            result[resultLength++] = unicode;
+        } else {
+            result.push_back(unicode);
         }
     }
 
-    if (resultLength > 0) {
+    if (result.size() > 0) {
         if (stringp) {
-            *stringp = [NSString stringWithCharacters:result length:resultLength];
+            *stringp = [NSString stringWithCharacters:result.data() length:result.size()];
         }
 
-        IwFree(result);
+        _scanLocation = i;
         return YES;
-    } else {
-        _location = oldLocation;
-
-        IwFree(result);
-        return NO;
     }
+
+    return NO;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (BOOL)scanDecimal:(NSDecimal*)decimalValue {
-    UNIMPLEMENTED();
-    return StubReturn();
+    // TODO #2369: This could probably be made more precise if it scanned directly from string -> NSDecimalNumber
+    // but WinObjC does not have [NSDecimalNumber decimalNumberWithString:] yet
+    double doubleValue;
+    if (![self scanDouble:&doubleValue]) {
+        return NO;
+    }
+
+    if (decimalValue) {
+        *decimalValue = [[NSDecimalNumber numberWithDouble:doubleValue] decimalValue];
+    }
+
+    return YES;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (BOOL)scanHexDouble:(double*)result {
```

Comment:
sscanf doesn't return where it stopped scanning. you can sort of work around that by specifying "%a%s" as the format, but that scans the whole rest of the string unnecessarily... you could also do a bunch of preprocessing to give it a prepped state, but then it just ends up being the same, but with sscanf instead of ldexp